modules/parent-join/src/main/java/org/elasticsearch/join/mapper/ParentIdFieldMapper.java

```diff
@@ -73,7 +74,9 @@ public boolean eagerGlobalOrdinals() {
 
         @Override
         public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
-            throw new UnsupportedOperationException("Cannot fetch values for internal field [" + typeName() + "].");
+            // Although this is an internal field, we return it in the list of all field types. So we
```

Comment:
This is not an ideal fix. I tried instead to hide this internal field type, so that it's not returned when we list all field types through `SearchExecutionContext`, but this would require a substantial refactor. Since it's an important bug that can prevent Kibana Discover from displaying fields when the index has a `join` field, I opted for this immediate straightforward fix.

Reply:
I'm wondering if it would be possible to extend the logic in

https://github.com/elastic/elasticsearch/blob/3bd594ebe86044c9b8280b75d68f7fe73512aacf/server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java#L76

to also skip in case of the `ParentIdFieldMapper` (perhaps it should also be a `MetadataFieldMapper`?)

Reply:
This is a good question. To me this field type didn't seem like a good fit for `MetadataFieldMapper`. Historically every metadata field has had a fixed name (like `_doc_count`, `_ignored`, etc.) and are meant to be rarely enabled/ configured by users.

This feels more like an 'internal' field. We have other examples where fields produce other fields as an implementation detail, like `flattened` and `_keyed`, and `text` and `_index_prefixes`... and these have caused problems too! Separately it'd be good to step back and think about how to better handle these internal fields.

Reply:
Thanks for sharing this historical info. I had similar thoughts about a new "internal" field category. In the context of the urgency of this PR I was mainly wondering whether instead of hacking the `ParentIdFieldMapper` class in this PR we would hack FieldFetcher to be aware of internal field names, and extend the code line I linked to (a first step would just be an `instanceof ParentIdFieldMapper` check there). This could later be more naturally expanded to a general "isInternalField" check as it's closer to the wildcard expansion step. I'm also fine going forward with the current state here if you prefer that.

Reply:
Unfortunately it's difficult to extend `FieldFetcher` in a targeted way like that, since many mappers including these join mappers live in modules. (This makes quick fixes harder, but is great overall because it encourages us to think through mapper design). I'm also reluctant to add a method like 'isInternal' to field mappers without thinking through the 'internal field' concept more deeply.

Reply:
ok, let's take this as first step then.